private/tools/java/rules/jvm/external/Hasher.java

```diff
@@ -0,0 +1,50 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package rules.jvm.external;
+
+import java.io.BufferedInputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+/** A tool to compute the sha256 hash of a file. */
+public class Hasher {
+
+  public static void main(String[] args) throws NoSuchAlgorithmException, IOException {
+    // Since this tool is for private usage, just do a simple assertion for the filename argument.
+    assert (args.length == 1) : "Please specify the path of the file to hash.";
+
+    String filename = args[0];
+    byte[] buffer = new byte[8192];
+    int count;
+    MessageDigest digest = MessageDigest.getInstance("SHA-256");
+    try (BufferedInputStream bufferedInputStream =
+        new BufferedInputStream(new FileInputStream(filename))) {
+      while ((count = bufferedInputStream.read(buffer)) > 0) {
+        digest.update(buffer, 0, count);
+      }
+    }
+    // Convert digest byte array to a hex string.
+    byte[] hashDigest = digest.digest();
+    StringBuffer hexString = new StringBuffer();
```

Comment:
pls don't use `StringBuffer`, all of its methods are `synchronized`, `StringBuilder` is what you want :)

See `StringBuffer` javadoc for details https://docs.oracle.com/javase/10/docs/api/java/lang/StringBuffer.html

Reply:
done.